rfcs/theme-shape.md

```diff
@@ -0,0 +1,228 @@
+# RFC: Update theme shape
+
+@miroslavstastny, @layershifter
+
+_This RFC updates [theme-tokens.md](./theme-tokens.md) RFC. Once this on is approved, we should merge the two RFCs together._
+
+## Summary
+
+As described in [theme-tokens.md](./theme-tokens.md) the current theme consists of global and alias tokens.
+All tokens are injected into DOM as CSS variables, resulting in 1200 CSS variables. With this approach, when a theme switches, there is no need to re-compute any component styles, by just setting the new token values to the CSS variables, all components change their appearance accordingly.
+
+However, 1200 CSS variables in DOM affects browser performance. This RFC proposes ways how to improve theme performance.
+
+## <a name="performance-analysis"></a>Performance analysis
+
+The current theme consists of ~1200 tokens which are injected into DOM as CSS variables. It impacts not only Javascript performance but also browser performance in styles computation phase.
+
+To see the performance impact we render 20 `FluentProvider` components side by side (**each** injecting a class with 1200 CSS variables) - mount and unmount 10 times and measure the performance in Chrome profiler with 6x CPU slowdown. After each mount, a CSS attribute is accessed to force reflow in order to be able to measure the rendering performance.
+
+### Master
+
+Current implementation in `@fluentui/react-theme@9.0.0-alpha.22`.
+
+- 1200 CSS variables in a theme = ~650 global tokens + ~550 alias tokens
+- alias token implemented as a "reference" to a global token: `--alias-token: var(--global-token)`.
+- **Each reflow takes ~150ms (1500 ms total for the 10 mounts)**
```

Comment:
How often are users changing themes? While noticable, 1.5 seconds doesn't seem like a long time to fully change themes.  
How good a job do browsers do with replacing CSS variables? Is the entire UI frozen or is it more progressive?

Reply:
Agree with the theme switching, I think we should not optimize for that as it is a rare operation.

But the problem is the reflows also happen during regular application lifetime, including **initial application launch**. It is application developers' responsibility to minimize the reflows, but we still should optimize the perf.

> How good a job do browsers do with replacing CSS variables?

That's something we still need to investigate.
Also another missing investigation - is only the reflow performance affected by the number (and structure - inline vs referencing globals) of CSS variables or does it impact any style computations anywhere inside the DOM subtree?